tutorial/7 - utils.test.ts

```diff
@@ -0,0 +1,197 @@
+ *
+ * If you see this variable, you should do something about it. :-)
+ */
+export const __YOUR_TURN__ = {} as any;
+
+// Silence TypeScript's import not used errors.
+expect(pairwise).toBe(pairwise);
+expect(scan).toBe(scan);
+expect(struct).toBe(struct);
+
+/**
+ * In the `sherlock-utils` lib, there are a couple of functions that can combine
+ * multiple values of a single `Derivable` or combine multiple `Derivable`s into
+ * one. We will show a couple of those here.
+ */
+describe.skip('utils', () => {
+    /**
+     * As the name suggests, `pairwise()` will call the given function with both
+     * the current and the previous state.
+     *
+     * *Note functions like `pairwise` and `scan` can be used with any callback,
+     * so it can be used both in a `.derive()` step and in a `.react()`*
+     */
+    it('pairwise', () => {
+        const myCounter$ = atom(1);
+        const reactSpy = jest.fn();
+
+        /**
+         * ** Your Turn **
+         *
+         * Now, use `pairwise()`, to subtract the previous value from the
+         * current.
+         *
+         * *Hint: check the overloads of pairwise if you're struggling with
+         * `oldVal`.*
+         */
+        myCounter$.derive(__YOUR_TURN__).react(reactSpy);
+
+        expect(reactSpy).toHaveBeenCalledExactlyOnceWith(1, expect.toBeFunction());
+
+        myCounter$.set(3);
+
+        expect(reactSpy).toHaveBeenCalledTimes(2);
+        expect(reactSpy).toHaveBeenCalledWith(2, expect.toBeFunction());
+
+        myCounter$.set(45);
+
+        expect(reactSpy).toHaveBeenCalledTimes(3);
+        expect(reactSpy).toHaveBeenCalledWith(42, expect.toBeFunction());
+    });
+
+    /**
+     * `scan` is the `Derivable` version of `Array.prototype.reduce`. It will be
+     * called with the current state and the last emitted value.
+     *
+     * (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)
+     *
+     * *Note: as with `pairwise()` this is useable in both a `.derive()` and
+     * `.react()` method*
+     */
+    it('scan', () => {
+        const myCounter$ = atom(1);
+        const reactSpy = jest.fn();
+
+        /**
+         * ** Your Turn **
+         *
+         * Now, use `scan()`, to add all the emitted values together
+         */
+        myCounter$.derive(__YOUR_TURN__).react(reactSpy);
+
+        expect(reactSpy).toHaveBeenCalledExactlyOnceWith(1, expect.toBeFunction());
+
+        myCounter$.set(3);
+
+        expect(reactSpy).toHaveBeenCalledTimes(2);
+        expect(reactSpy).toHaveBeenCalledWith(4, expect.toBeFunction());
+
+        myCounter$.set(45);
+
+        expect(reactSpy).toHaveBeenCalledTimes(3);
+        expect(reactSpy).toHaveBeenCalledWith(49, expect.toBeFunction());
+
+        /**
+         * *BONUS: Try using `scan()` (or `pairwise()`) directly in the
+         * `.react()` method.*
+         */
+    });
+
+    it.skip('pairwise - BONUS', () => {
+        const myCounter$ = atom(1);
```

Comment:
Just for the record: don't suggest to replace all these "copy the same implementation here" by a variable. The whole point of this is to show that using a variable works but not using a variable fails!